torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -728,11 +735,21 @@ Tensor prod_backward(
     return grad;
   }
   dim = at::maybe_wrap_dim(dim, input.sizes().size());
-  if (!keepdim && input.dim() != 1) {
+  if (!keepdim) {
+    // `prod` reduces the dimension at `dim`,
+    // so, unsqueeze `grad` and `result` at dim.
     grad = grad.unsqueeze(dim);
     result = result.unsqueeze(dim);
   }
-  if (is_meta_in_composite_kernels(input)) {
+  if (input.numel() == 0) {
+    // When input has a zero sized dimension (empty tensor),
+    // we don't need to actually compute the grads.
+    // So we just reshape `grad` as `input`.
+    return grad.expand_as(input);
+  }
```

Comment:
Okay I thought about it some more, this LGTM. In particular, `prod_safe_zeros_backward` does the same thing (propagate the requires_grad-ness of the `grad` Tensor).

Reply:
NB: This is not actually correct. Let's say we have input of shape (3, 0), and did `out = input.prod(dim=1)`.

`grad` would have shape [3]; `grad.expand_as(input)` would raise an error

Reply:
This is going to sound really stupid, but what if we just did `grad * (result / input).conj();` ? That's what happens in the zero-numel case in the nonzero() path as well.

Reply:
This works because of the following code before the call to `grad.expand_as(input)`.
```
if (!keepdim) {
     // `prod` reduces the dimension at `dim`,
     // so, unsqueeze `grad` and `result` at dim.
     grad = grad.unsqueeze(dim);
     result = result.unsqueeze(dim);
}
```

Reply:
Oh I see, you are right. Thanks for adding the new sample